### PR TITLE
V20 Fixes and changes for Streaming

### LINKED ISSUE #26702: Make Kafka topics configurable
The suffixes for names of topics to which Mantid subscribes for data from Kafka are hardcoded in `KafkaTopicSubscriber.cpp`. The instrument name is used as the prefix. This has worked fine for ISIS but it is clear for the ESS we now need to support multiple event and sample environment topics and allow the names to be configured. Topic names should be configured through the Facilities file so that the user is not faced with them.

Detector IDs and sample environment source names are unique so consuming from multiple topics should be straightforward; nothing will need to change in the way the messages are processed.

The Kafka listener should continue to work as before for ISIS instruments using it: ZOOM, MARI and LARMOR, so their entries in the Facilities file should be updated to contain the existing topic names.

The `StartLiveData`/`KafkaLIstener` documentation should be updated to document where the topic names are configured.

_Ideally this would be working for December 2019 beamtime at V20_

### PR DESCRIPTION
**Pull request by LamarMoore**
**Description of work.**

This work largely comprises target kafka development undertaken at the V20 beamline to facilitate streaming of live data for ESS.

**To test:**

Fixes #26702. in addition to other changes related to issues discovered making this change. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by matthew-d-jones** [comment]
@DanNixon @LamarMoore 
There will be another schema update soon: https://github.com/ess-dmsc/streaming-data-types/pull/45
It should very little work to move to from the one here though. There will be some additional fields in the run start message, however these are for the benefit of our file writer software and Mantid can ignore them. The main (maybe only) change which concerns Mantid will be that I have got rid of the separate detector-spectrum map message and instead added it as an optional field in the run start message, I think this will allow some simplification of the Mantid listener.

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp [hunk 1]
```diff
@@ -244,19 +242,18 @@ void KafkaHistoStreamDecoder::initLocalCaches(
...
-    if (!instName.empty()) {
-      loadInstrument<DataObjects::Workspace2D>(instName, histoBuffer,
-                                               jsonGeometry);
-      if (rawMsgBuffer.empty()) {
-        histoBuffer->rebuildSpectraMapping();
-      }
-    } else {
-      g_log.warning(
-          "Empty instrument name received. Continuing without instrument");
+  // Load the instrument if possible but continue if we can't
```
Comment (DanNixon) [inline Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: Should these changes also go into the event decoder?

**Review by DanNixon — CHANGES_REQUESTED** [review]
This is just what I found from a code review.

File: Framework/LiveData/src/Kafka/KafkaTopicSubscriber.cpp [hunk 2]
```diff
@@ -20,28 +21,35 @@ using RdKafka::TopicMetadata;
 
 namespace {
 /// Timeout for message consume
-const int CONSUME_TIMEOUT_MS = 30000;
-const std::string MAX_MESSAGE_SIZE = "100000000";
+const int CONSUME_TIMEOUT_MS = 300000;
+const std::string MAX_MESSAGE_SIZE = "200000000";
+// Receive buffer size must be at least MAX_MESSAGE_SIZE+512
+const std::string RECEIVE_BUFFER_SIZE = "200000512";
```
[the remark this answers is not in the document]
Comment (LamarMoore) [reply Framework/LiveData/src/Kafka/KafkaTopicSubscriber.cpp]: We needed the message size this large to cope with the LOKI geometry I think. Should be fine for V20 though.

File: Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp (hunk 1 above)
Comment (LamarMoore) [reply Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: Aren't these changes in the event decoder as well? (provided we are talking about the same thing)

Comment (DanNixon) [reply Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: They don't seem to be. The buffer creation code is untouched.

Comment (LamarMoore) [reply Framework/LiveData/src/Kafka/KafkaHistoStreamDecoder.cpp]: They look almost identical (side-by-side) except for the closing brace after the buffer creation `599` in the event decoder and `245` in the histo decoder. I will make that change and resubmit.

**Review by DanNixon — APPROVED** [review]
:+1: Looks good to be now.

**Pull request merged**